### DISCUSSION TITLE
Dynamic Label placement for features depending on available space

### DISCUSSION THREAD

**Discussion by AlbrechtWM** [openlayers/openlayers#16084] [General]
So I had another question I wanted to ask, this time about how to dynamically place labels for a given point/feature should there be sufficient space near the point to place them where the label will not overlap with another label or point.

It did not appear there was any "out of the box" functionality to do this (correct me if I am wrong).  I am happy to write my own approach for this, but did not know how to extract "hit detection" for a label.  Obviously, it exists as declutter knows when to hide or show a label.  But I am trying to move the label around the point to a place where it can fit if such a space exists.

I am aware this can pose major performance issues given the nature of the problem, but I am perfectly happy to optimize that aspect of it away (if I can).

Any general guidance on approach or existing functions that I could use would be appreciated.  

Thanks!

**Comment by mike-000** [comment]
See https://github.com/openlayers/openlayers/issues/11449#issuecomment-1593438241

**Reply by AlbrechtWM** [reply]
Mike, thanks I had seen this in the past and have used it for inspiration.  I am having some success so far and will attempt to upload an image when github actually allows that to take place (keeps failing).  Basically, I am attempting to place labels in the same 8 radial directions (Top/North, TopRight/NorthEast, etc).  For the moment, I am just using rectangles, not text, as the size is fixed and the hit detection is easy to do.

My problem is the variable nature of Text, and essentially, how do I manage to get the *size* of the Text *before* it's placed on the screen.  I saw Text.getScale() existed but so far returns undefined when I created some ol/style/Text objects.  In my scenario, I am not labelling countries, or anything with a fixed, predetermined location.  The data is dynamic, with a lot of datapoints, with a requirement that the text not collide with any other text, or any of the points.  Fortunately, another requirement is to not bother trying to place dynamic labels until a certain zoom has been reached, and also, only a subset of all points (the most "important" points) need to have labels.

Is it possible to get the size (in screen size) of an ol/style/Text object before it has been added to the map or rendered in some way?

**Comment by mike-000** [comment]
It will work equally well for points, if you wish to display a point symbol with text in one of eight possible positions around it the styles will declutter each other so no more than one is displayed per feature.  The application should not need to know the size of the text, decluttering will deal with that.  Use zIndex to ensure high importance features get priority.  You could add a layer with only point symbols to show features where labels cannot be shown.  A style function takes a resolution so you can change style depending on zoom level https://stackblitz.com/edit/js-epxqeb?file=package.json,index.html,index.js

**Reply by AlbrechtWM** [reply]
Mike,

  Fascinating.  Just looking at this example you linked, and I might be able to potentially modify it to suit the research/testing I am doing (I am dealing with a large legacy application so figuring out where / when I can make changes is tricky).  The only caveat I have left is that the points themselves *must never* be decluttered, but the labels can be decluttered provided they give a "best effort" to try to display in some place.  It looks like your labels are decluttering/dynamically placing perfectly, and I also really like the fade being applied to "less important" points if that's indeed what's going on.  My experiments with zIndex stuff just resulted in drawing directly on top of other things, or when applied with declutter, caused the higher priority points to have a higher likelihood of showing but not a guarantee due to the vagaries of the rTree being employed.

  So yeah, if there is a way to have the points never decluttered, but the labels decluttered (and still respect non-decluttering points when doing so) this will work.

  Appreciate your replies!

**Comment by mike-000** [comment]
You could experiment with using `declutterMode` in the image and text components of the style.

**Reply by AlbrechtWM** [reply]
Mike,

  Yeah I have tried playing with declutterMode in the styles before (declutter, obstacle, none), but unfortunately I do not think it actually works.  The problem is the VectorLayer has to have a declutter value (it will default to false if you don't specify).  In addition, whatever the Vector Layers declutter setting is overrides any style-based declutterMode as far as I can tell (perhaps I should add that as a bug report for @ahocevar ).

  Even still, if the requirements (unusual as I know they must seem) are points are to always be shown (no declutter) but labels can be declutter, I don't think it's possible to declutter with other things that are not set to decluttter.

(Although, I *could* create invisible points that overlap the actual uncluttered points and declutter against them?)

Anwyays, appreciate the ideas and replies.  I will keep cranking away =)

**Reply by ahocevar** [reply]
@AlbrechtWM `declutterMode` has been fixed somewhere in v9.x. And decluttering (including style specific `declutterMode` settings) *only* takes place when the layer's `declutter` option is set.

**Reply by AlbrechtWM** [reply]
@ahocevar Thanks ahocevar.  I am on v9.2.4, but haven't needed to leverage declutterMode yet in the styles so I will let that lie for the moment.  If it comes up il try to do some further testing on it.
